Add new microengine utilities

zv opened the pull request:
- Exceptions for common conditions encountered by engines
- Method to convert Unix paths to Windows paths suitable for use by WINE engines
- `content_type`, widely duplicated logic for determining the likely file type extension of some bytes(extracted from `filseclab`)
- `EngineInfo` class, a standard object for storing duplicate scanner information (NEED FEEDBACK)
- `collect_scan`, a decorator which automatically handles errors and boilerplate scanner metadata (NEEDS FEEDBACK)
- `create_scanner_exec`, a simple scanner-command runner which automatically terminates after a timeout
- `ArtifactTempfile`, a cut-down version of `AsyncArtifactTempfile` which works in both asynchronous and traditional context manager. No longer allows manipulation of the underlying file object.

mrtizmoatwork commented on src/microengine_utils/errors.py:
```diff
@@ -0,0 +1,88 @@
+
+
+class ScannerNotFoundError(BaseScanError):
+    """Scanner couldn't find signature / models"""
```
I think the comment here is incorrect? This comment doesn't match the class name. And this comment is a duplicate of line 57.

mrtizmoatwork commented on src/microengine_utils/fs.py:
```diff
@@ -0,0 +1,143 @@
+import asyncio
```
this file is named "fs.py", what does "fs" mean? Please use full words in the file name.

mrtizmoatwork commented on src/microengine_utils/fs.py:
```diff
@@ -0,0 +1,143 @@
+        return await self._executor(self.__exit__, exc, value, tb)
+
+
+try:
```
do we want to do this in every file for every engine? Many engines don't care about the actual file name. Maybe make this an optional thing.

mrtizmoatwork commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+
+
+def collect_scan(verbose: 'bool' = False) -> 'Callable[..., ScanResult]':
+    """Decorator for `async_scan` to automatically handle errors and boilerplate scanner metadata
```
if this is for `async_scan` should we call it `collect_async_scan`? And should we have a similar function for the `sync_scan` ?

Rizato replied:
Since this is a decorator, I think it should check if the function is a coroutine, and act based on that.

mrtizmoatwork commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+                    if verbose:
+                        statsd.increment(SCAN_NO_RESULT, tags=tags)
+                # Verdict reported
+                elif scan.bit is True and (scan.verdict is True or scan.verdict is False):
```
the right side of the `and` has a test for both True and False, what other options are there? Is `None` an option for that value?

Rizato replied:
`verdict` defaults to `False`

Rizato replied:
Though I think we were discussing a move to allow bit True, verdict None to create a suspicious label.

mrtizmoatwork commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+            def update(...):
+                update: Mapping[str, str] = ... #
+                self.info.signature_version = update['definitions_version']
+                self.info.engine_version = update['engine_version']
```
`engine_version` is not defined in `__init__`. Should it be?

Rizato commented on src/microengine_utils/fs.py:
```diff
@@ -0,0 +1,143 @@
+
+def path_template(tmpl):
+    if platform.system() == 'Windows':
+        MICROENGINE_INSTALL_ROOT = getenv('PSC_INSTALL_DIR', 'C:\\microengine\\')
```
import os directly, and use `os.getenv`

Rizato commented on src/microengine_utils/fs.py:
```diff
@@ -0,0 +1,143 @@
+            os.close(fd)
+
+        self.blob = blob
+        self._executor = asyncio.get_event_loop().run_in_executor
```
This will block anyone that tries to init a file in an async context.

zv replied:
good call, I moved this logic out into `__enter__` to prevent this

Rizato commented on src/microengine_utils/fs.py:
```diff
@@ -0,0 +1,143 @@
+            getattr(os, 'O_SEQUENTIAL', 0)  # optimize for (but don't require) sequential access
+        )
+
+        fd = os.open(self.name, flags, 0o666)
```
Use a preexisting variable for `0o666` or define one

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+            stderr=stderr,
+            stdin=stdin,
+        )
+        streams = await asyncio.wait_for(proc.communicate(), timeout=timeout)
```
The entire scan call is wrapped in an `asyncio.wait_for` based on the bounty duration so this isn't necessary. 

Leave in the asyncio.TimeoutError handling though.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+
+
+def collect_scan(verbose: 'bool' = False) -> 'Callable[..., ScanResult]':
+    """Decorator for `async_scan` to automatically handle errors and boilerplate scanner metadata
```
```suggestion
    """Decorator for `scan_async` to automatically handle errors and boilerplate scanner metadata
```

mrtizmoatwork replied:
@zv please create another copy of this function that works for `sync` and have a different name that make it clear it is for sync vs async

Rizato replied:
I think it looks good now. Just need to fix the comment.

 It's a pretty common pattern for a decorator that handles both sync and async, since the driver functions the decorator creates are different.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+        return driver
+
+
+class EngineInfo:
+    """A standard object to store scanner & signature metadata
+
+    Notes::
+
+    Some engines report signature metadata with their scan's output, others
+    do so during expensive signature updates.
+
+    You can use `EngineInfo` in both cases, providing a way to store the
+    results of `update` or just to simplify the logic of setting up
+    a `ScanResult`'s scanner info ::
+
+        class Engine(Scanner):
+            info = EngineInfo(version=polyswarm_nanoav.__version__)
+
+            def update(...):
+                update: Mapping[str, str] = ... #
+                self.info.signature_version = update['definitions_version']
+                self.info.engine_version = update['engine_version']
+
+            def sync_scan(...)
+                scan_result: ScanResult = do_scan(...)
+                self.info.update_verdict(scan_result.verdict)
+                return scan_result
+
+    """
+    def __init__(self):
+        self.version = version
+        self.operating_system = None
+        self.architecture = None
+        self.signatures_version = None
+        self.vendor_version = None
+
+    def dict(self, include_none=False):
+        return {k: v for k, v in self.__dict__.items() if (include_none or v is not None)}
```
If this is intended to be used to input data into an existing `Verdict` metadata, the functionality should be included as part of the class. 

Also, another option is to add a method the creates a prefilled Verdict metadata that just needs a malware_family.

Rizato replied:
And, may be better as a part of polyswarm-artifact

Rizato replied:
I could see this being a part of polyswarm-client though too.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+                    verdict=False,
+                    metadata=Verdict().set_malware_name('').set_extra('scan_error', e.event_name),
+                )
+            else:
```
From my understanding of try else, this won't be called when there is an exception. 

That means exceptions during scans won't be reported to datadog.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+
+        return driver
+
+
```
Need to return wrapper here as well.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+                    # boilerplate data (e.g `platform`, `machine`, `signature_info`, etc.)
+                    if isinstance(getattr(self, 'scanner_metadata', None), EngineInfo):
+                        scan.metadata.set_scanner(**self.info.dict())
+                    scan.metadata = scan.metadata.json()
```
`scan.metadata` is a string. This is due to the contract which only accepts a string. We could change this in psc to accept a Verdict object instead of a string, but that will break compatibility with the contract. That may be a path we move, but we should make that change first. 

You could try to convert the string into a Verdict object, and then modify it, but I think there is a better answer. 

This functionality belongs in the EngineInfo object, and should add the data to a `Verdict` before it is passed to a `ScanResult`

Rizato replied:
To reinforce why this belongs in `EngineInfo`, not everyone is going to be using datadog stats, but they should be able to use `EngineInfo` regardless.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,165 @@
+
+                return scan
+
+        return driver
```
I would refactor this a bit. 

I think everything in the else block should come after finally, and not be in an else block at all. 

The scan time we generate we get should be the time it takes to scan, not the time it takes to scan and report results to datadog.

Rizato commented:
I am definitely a fan of the ideas in here. 

I see potential for EngineInfo. I think we could still use it entirely in a decorator even with my suggestions about moving the functinoality to create a Verdict in the class. 

Say we add the ability to set only the malware family on a ScanResult. The decorator could act if malware_family is set and create the full metadata object behind the scenes.

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,213 @@
+            finally:
+                # If engines define `scanner_metadata`, we'll automatically include all of the
+                # boilerplate data (e.g `platform`, `machine`, `signature_info`, etc.)
+                if isinstance(getattr(self, 'scanner_metadata', None), EngineInfo):
```
I'm not a fan of hardcoding a string for a variable the person has to add to their `Scanner`

We can discuss other options in the call.

zv commented:
@Rizato @mrtizmoatwork I believe I've addressed all your comments & added unit tests to verify new code's functionality.

I think it's still important that you guys give this another review, firstly to make sure you don't see any issues with e.g exporting some functions from the pkg root rather than submodules, how `scanalytics` works, etc. On my end, I would love another set of eyes because I have that sensation that something essential is being left out of `EngineInfo` and the constants declaration.

@sbnetops, @mrtizmoatwork mentioned you thought there was an issue with creating a fresh event loop in a synchronous function? I wasn't able to find any issues in unit tests and doing so would allow us to maintain a single function, rather than two duplicates which differ only in their function type qualifier.

Rizato commented:
One reason to do multiple functions is that by the time `scan_sync` is called we are in synchronous execution. The an async driver function won’t work correctly.

mrtizmoatwork commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,139 @@
+from .errors import BaseScanError, CalledProcessScanError
+
+
+async def create_scanner_exec(
```
we should have a sync version of this too

Rizato commented on src/microengine_utils/scanner.py:
```diff
@@ -0,0 +1,148 @@
+        def collect_error(err: 'BaseScanError', tags: 'List[str]') -> 'ScanResult':
+            # If we encountered a scan error, we still return a scan result with `scan_error`
+            # documenting the error which occurred
+            v = Verdict().set_malware_family('').add_extra('scan_error', err.event_name)
```
Since bit is false this metadata won't be read anywhere

zv replied:
This is still really useful, even if we don't end up using the metadata. I have already found it really helpful in debugging scan tests, which now report _why_ they are failing, rather than a naked `ScanResult`.

Rizato commented on src/microengine_utils/config.py:
```diff
@@ -0,0 +1,115 @@
+            elif field.alias in kwargs:
+                setattr(self, name, kwargs[field.alias])
+
+    class Config:
+        allow_mutation = True
+        allow_population_by_field_name = True
```
What is this for?

zv replied:
This allows you to create or update an `EngineInfo` instance by _either_ it's `alias`, so it can be directly populated from field names from `set_scanner` (e.x `vendor_version` & `signatures_version`) **or** by it's "true" attribute names (`engine_version` & `definitions_version`)

mrtizmoatwork commented on src/microengine_utils/filesystem.py:
```diff
@@ -0,0 +1,117 @@
+try:
+    import puremagic
+
+    async def content_type(content: 'Union[bytes, bytearray]') -> 'Optional[str]':
```
@Rizato @zv will this work ok in a sync context?

Rizato replied:
I assume sync users would call puremagic directly.

Rizato replied:
We may want to change the name to reflect the sync function it wraps

Rizato replied:
Why is this function declared in a try-except, though? It's weird.  I would prefer the function raise an exception if puremagic isn't available.

zv replied:
puremagic isn't required by every engine

Rizato commented on src/microengine_utils/filesystem.py:
```diff
@@ -0,0 +1,117 @@
+    return PureWindowsPath('Z:\\').joinpath(*(p.replace('/', '\\') for p in rest))
+
+
+async def winepath(path: 'os.PathLike', output='Windows') -> 'PureWindowsPath':
```
We should have a sync version of this one, though.

zv replied:
I was having questions about if we should even include this at all, it's only useful to WINE engines and so far none are using synchronous scanning. That said, there's something to be said for `microengine-utils` having a single definitive implementation of `winepath`.

I think we should include it as-is for future one offs, or just take it out entirely. It's up to you though.
